clients/src/main/java/org/apache/kafka/common/requests/OffsetForLeaderEpochRequest.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.requests;
+
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.protocol.types.Struct;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class OffsetForLeaderEpochRequest extends AbstractRequest {
+    public static final String TOPICS = "topics";
+    public static final String TOPIC = "topic";
+    public static final String PARTITIONS = "partitions";
+    public static final String PARTITION_ID = "partition_id";
+    public static final String LEADER_EPOCH = "leader_epoch";
+
+    private Map<String, List<Epoch>> epochsByTopic;
```

Comment:
This was a pretty big change, but the final one of your first round of comments. Committed now.